ulib/experimental/Steel.SelEffect.fsti

```diff
@@ -361,3 +400,103 @@ val write (#a:Type0) (r:ref a) (x:a) : SteelSel unit
 let sel (#a:Type) (#p:vprop) (r:ref a)
   (h:rmem p{FStar.Tactics.with_tactic selector_tactic (can_be_split p (vptr r) /\ True)})
   = h (vptr r)
+
+val intro_vrefine
+  (v: vprop) (p: (normal (t_of v) -> Tot prop))
+: SteelSel unit v (fun _ -> vrefine v p)
+  (requires (fun h -> p (h v)))
+  (ensures (fun h _ h' -> normal (h' (vrefine v p) == h v)))
+
+val elim_vrefine
+  (v: vprop) (p: (normal (t_of v) -> Tot prop))
+: SteelSel unit (vrefine v p) (fun _ -> v)
+  (requires (fun _ -> True))
+  (ensures (fun h _ h' -> normal (h' v == h (vrefine v p))))
+
+val intro_vdep
+  (v: vprop)
+  (q: vprop)
+  (p: (t_of v -> Tot vprop))
+: SteelSel unit
+    (v `star` q)
+    (fun _ -> vdep v p)
+    (requires (fun h -> q == p (fst (h (v `star` q)))))
+    (ensures (fun h _ h' ->
+      let x1 = h (v `star` q) in
+      let x2 = h' (vdep v p) in
+      q == p (fst x1) /\
+      dfst x2 == fst x1 /\
+      dsnd x2 == snd x1
+    ))
+
+val elim_vdep
+  (v: vprop)
+  (p: (t_of v -> Tot vprop))
+  (q: vprop)
+: SteelSel unit
+  (vdep v p)
+  (fun _ -> v `star` q)
+  (requires (fun h -> q == p (dfst (h (vdep v p)))))
+  (ensures (fun h _ h' ->
+      let x1 = h' (v `star` q) in
+      let x2 = h (vdep v p) in
+      q == p (fst x1) /\
+      dfst x2 == fst x1 /\
+      dsnd x2 == snd x1
+  ))
+
+let elim_vdep2
+  (v: vprop)
+  (p: (t_of v -> Tot vprop))
+  (q: vprop)
+: SteelSel (Ghost.erased (t_of v))
+  (vdep v p)
+  (fun _ -> v `star` q)
+  (requires (fun h -> q == p (dfst (h (vdep v p)))))
+  (ensures (fun h res h' ->
+      let x1 = h' (v `star` q) in
+      let x2 = h (vdep v p) in
+      q == p (fst x1) /\
+      dfst x2 == fst x1 /\
+      dsnd x2 == snd x1 /\
+      Ghost.reveal res == fst x1
+  ))
+=
+  elim_vdep v p q;
+  let r = gget (v `star` q) in
+  Ghost.hide (fst (Ghost.reveal r))
+
+(* FIXME: WHY WHY WHY does this version fail:
+let elim_vdep3
+  (v: vprop)
+  (p: (t_of v -> Tot vprop))
+  (q: vprop)
+: SteelSel (Ghost.erased (t_of v))
+  (vdep v p)
+  (fun _ -> v `star` q)
+  (requires (fun h -> q == p (dfst (h (vdep v p)))))
+  (ensures (fun h res h' ->
+      let x1 = h' (v `star` q) in
+      let x2 = h (vdep v p) in
+      q == p (fst x1) /\
+      dfst x2 == fst x1 /\
+      dsnd x2 == snd x1 /\
+      Ghost.reveal res == fst x1
+  ))
+=
+  let r = gget (vdep v p) in
+  elim_vdep v p q; // FAIL here: cannot prove precondition
+  Ghost.hide (dfst #(t_of v) #(vdep_payload v p) (Ghost.reveal r))
+*)
```

Comment:
I am (separately) trying to write unit tests for vdep, but already in this simple case I hit a problem maintaining equalities when performing no-op stateful operations before `elim_vdep`. @R1kM Can you please help me find out what I am doing wrong? Thanks in advance!

Reply:
A part of the problem was related to equalities missing in frame_equalities, an issue you hit previously, but the fix I proposed wasn't sufficient. For now, I reenabled quadratic equality generation, and opened #2280 to keep track of this problem.
The other part of the problem was related to the specification expressed on a composite vprop, (v * q). Lemmas relating sel (v * q) to sel v, sel q are (on purpose) not available to the SMT solver. As such, I modified your spec to be on atomic subresources, which allowed verification to succeed.

I also fixed an overly eager normalization on the types of vdep.

Unrelated to this issue, could we put these examples/unit tests in a separate file, rather than sticking them in Steel.SelEffect? I think ultimately we should have separate files for the different components of Steel.SelEffect, with a top-level empty files including all others (similar to FStar.Tactics). Let's do this refactor later, I'll add it to #2243

Reply:
Thank you for your fixes, and for re-enabling quadratic frame_equalities for now. This allowed me to push a more meaningful unit test, namely singly linked lists, see below.

Of course I didn't mean `elim_vdep3` to be a unit test by itself, but I had to be able to define it, or more precisely a consolidated version of then-`elim_vdep2` and `elim_vdep3`, to carry on my singly linked list test.